### ISSUE TITLE
Virtual Machine detection in Genshin Impact (VMProtect)

### ISSUE THREAD

**Issue by strongtz** [ptitSeb/box64#914]
I'm trying to run Genshin Impact on Arch Linux ARM, which is running on a Snapdragon 888 SoC with Linux kernel v6.5-rc3. However, it will just say "*Sorry, this application cannot run under a Virtual Machine*". 

This particular wine build has also been tested on my desktop, and it can start genshin just fine.

After some search, I figured that the **VMProtect** being used in the game is causing the issue. Anything else using VMProtect [like this one](https://notabug.org/Krock/dawn/issues/12#issuecomment-27405) also pops out the exact same error dialog.

I'd be glad if anyone could help. Thanks in advance :P

**Comment by strongtz** [comment]
Well, I finally found the code that is reporting *virtual machine* in [VMProtect source](https://github.com/classic130/VMProtect-Source/blob/master/runtime/core.cc#L287-L309)

```c++
	uint64_t val;
	uint8_t mem_val;
	__try {
		// set T flag
		__writeeflags(__readeflags() | 0x100);
		 val = __rdtsc();
		 __nop();
		 loader_data->set_is_debugger_detected(true);
	} __except(mem_val = *static_cast<uint8_t *>((GetExceptionInformation())->ExceptionRecord->ExceptionAddress), EXCEPTION_EXECUTE_HANDLER) {
		if (mem_val != 0x90)
			return true;
	}

	__try {
		// set T flag
		__writeeflags(__readeflags() | 0x100);
		__cpuid(cpu_info, 1);
		__nop();
		loader_data->set_is_debugger_detected(true);
	} __except(mem_val = *static_cast<uint8_t *>((GetExceptionInformation())->ExceptionRecord->ExceptionAddress), EXCEPTION_EXECUTE_HANDLER) {
		if (mem_val != 0x90)
			return true;
	}
```

@ptitSeb Sorry for disturbing, but would you possibly have any idea on this? :D

**Comment by ptitSeb** [comment]
Mmm, I should have handled this already with https://github.com/ptitSeb/box64/commit/d2547672f314d70470a2370ee8ff391a5c0b93d0. Looks like I have messed up something. I'll double-check what I have done.

**Comment by strongtz** [comment]
emm, any updates on this? :P

**Comment by strongtz** [comment]
still not working as of 8e7afcc5fb6c320126c06bc46f7691fb2199634d

**Comment by shenmo7192** [comment]
keep watching

**Comment by ptitSeb** [comment]
Mmmm, I see tgis game is free on Epic, and I used Heroix to launch it on linux with box64. But the game wants to download 67GB of data :( That's a lot for a quick test. When does this anti debugger message appear exactly?

**Comment by strongtz** [comment]
It appears when launching the executable under `Genshin Impact Game`, after download and unpacking completes

**Comment by strongtz** [comment]
As I mentioned previously though, [the test program here](https://notabug.org/Krock/dawn/issues/12#issuecomment-27405) should have the same problem.

The game executable is also here for convenience, not sure if it would launch without data.
[YuanShen.zip](https://github.com/ptitSeb/box64/files/12491874/YuanShen.zip)

**Comment by ptitSeb** [comment]
Are you sure this game use the method you mentionned for it's anti-debugg detection? because there are more than 1 way to do that.

**Comment by strongtz** [comment]
I'm only sure that it does use VMProtect

**Comment by shenmo7192** [comment]
> Are you sure this game use the method you mentionned for it's anti-debugg detection? because there are more than 1 way to do that.

he upload the game origin file. this may help.

**Comment by ptitSeb** [comment]
the exe needs a unity dll... and probably many other... :(  I'll try to download the whole damn thing.

**Comment by shenmo7192** [comment]
> the exe needs a unity dll... and probably many other... :( I'll try to download the whole damn thing.

I have the entire game. Let me try to make a minimium runable pack

**Comment by shenmo7192** [comment]
![图片](https://github.com/ptitSeb/box64/assets/47873776/5c864f97-0794-4344-a8ce-b60ad14b6545)
Yuanshen.exe+Unityplayer.dll, the minimum runable pack is here
[Desktop.zip](https://github.com/ptitSeb/box64/files/12502704/Desktop.zip)

**Comment by shenmo7192** [comment]
> As I mentioned previously though, [the test program here](https://notabug.org/Krock/dawn/issues/12#issuecomment-27405) should have the same problem.
> 
> The game executable is also here for convenience, not sure if it would launch without data. [YuanShen.zip](https://github.com/ptitSeb/box64/files/12491874/YuanShen.zip)

Kokomi is so cuteeeeeeeee!

**Comment by ptitSeb** [comment]
Ok, I reproduce the issue with the minimal package, good.

**Comment by ptitSeb** [comment]
Ok, I think I understand what was wrong with my current handling of the TF flag. Detection should pass now with latest box64.

**Comment by shenmo7192** [comment]
> I'm only sure that it does use VMProtect

I can now enter the white 原神，启动（Genshin Impact Start）screen, can see the huge mihoyo logo, then the logo fade out, then screen turn completely white....Can not see the Genshin Logo and can not use my cursor.....

**Comment by ptitSeb** [comment]
Well, that means this ticket is solved.
The other issues with the game are other issue now. And I'll have to download the full game to debug.

**Comment by shenmo7192** [comment]
> Well, that means this ticket is solved. The other issues with the game are other issue now. And I'll have to download the full game to debug.

Yeah you are right.

Thank you so much for fix the vm detect~

**Comment by strongtz** [comment]
Closing the issue thanks to the fix in eb1e607.

A new issue for Genshin Impact is here: #967
